libvast/vast/system/query_processor.hpp

```diff
@@ -0,0 +1,162 @@
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+
+#include <caf/behavior.hpp>
+#include <caf/fwd.hpp>
+
+#include "vast/fwd.hpp"
+#include "vast/uuid.hpp"
+
+namespace vast::system {
+
+/// A query processor takes (1) a query, (2) a processing step function,
+/// and (3) a predicate for implementing the following state machine.
+///
+/// ```
+///                    +----------------+
+///                    |                |
+///               +--->+      idle      |
+///               |    |                |
+///               |    +-------+--------+
+///               |            |
+///               |            | (run)
+///               |            v
+///               |    +-------+--------+
+///               |    |                |
+///               |    | await query id |
+///               |    |                |
+///               |    +-------+--------+
+///               |            |
+///               |            | (query_id, scheduled, total)
+///               |            |
+///               |            |      +------+
+///               |            |      |      |
+///               |            v      v      | (ids)
+///               |    +-------+------+-+    |
+///               |    |                +----+
+///               |    |  collect hits  |
+///               |    |                +<---+
+///               |    +-------+--------+    |
+///               |            |             |
+///               |            | (done)      |
+///               |            v             |
+///               |       XXXXXXXXXXXX       |
+///               |      XX request  XX      |
+///               +----+XX    more    XX+----+
+///                no    XX   hits?  XX   yes
+///                       XXXXXXXXXXXX
+/// ```
+
+class query_processor {
+public:
+  // -- member types -----------------------------------------------------------
+
+  enum state_name {
+    idle,
+    await_query_id,
+    collect_hits,
+  };
+
+  static constexpr size_t num_states = 3;
+
+  // -- constants --------------------------------------------------------------
+
+  /// Human-readable actor name for logging output.
+  static inline constexpr const char* name = "query-processor";
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  /// @warning Calls `set_default_handler(caf::skip)`.
+  query_processor(caf::event_based_actor* self);
```

Comment:
I'd prefer to pass the index handle here instead of with `start()`.

Reply:
That's not possible, unfortunately. This constructor is called by the `stateful_actor<...>` and we have the option of either taking no arguments or the `self` pointer.